Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,282 @@
+      _ openStreams: Int,
+      _ signalEndConnection: AsyncStream<Void>.Continuation
+    )
+    case closed(
+      _ openStreams: Int,
+      _ signalEndContinuation: AsyncStream<Void>.Continuation?
+    )
+  }
+
+  public typealias Inbound = RPCAsyncSequence<RPCResponsePart>
+  public typealias Outbound = RPCWriter<RPCRequestPart>.Closable
+
+  public let retryThrottle: RetryThrottle
+
+  private let executionConfigurations: ClientRPCExecutionConfigurationCollection
+  private let state: LockedValueBox<State>
+
+  public init(
+    server: InProcessServerTransport,
+    executionConfigurations: ClientRPCExecutionConfigurationCollection
+  ) {
+    self.retryThrottle = RetryThrottle(maximumTokens: 10, tokenRatio: 0.1)
+    self.executionConfigurations = executionConfigurations
+    self.state = LockedValueBox(.unconnected(server, []))
+  }
+
+  /// Establish and maintain a connection to the remote destination.
+  ///
+  /// Maintains a long-lived connection, or set of connections, to a remote destination.
+  /// Connections may be added or removed over time as required by the implementation and the
+  /// demand for streams by the client.
+  ///
+  /// Implementations of this function will typically create a long-lived task group which
+  /// maintains connections. The function exits when all open streams have been closed and new connections
+  /// are no longer required by the caller who signals this by calling ``close()``, or by cancelling the
+  /// task this function runs in.
+  ///
+  /// - Parameter lazily: This parameter is ignored in this implementation.
+  public func connect(lazily: Bool) async throws {
+    let (stream, continuation) = AsyncStream<Void>.makeStream()
+    try self.state.withLockedValue { state in
+      switch state {
+      case .unconnected(let server, let pendingStreams):
+        state = .connected(server, 0, continuation)
+        for pendingStream in pendingStreams {
+          pendingStream.finish()
+        }
+      case .connected:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Already connected to server."
+        )
+      case .closed:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Can't connect to server, transport is closed."
+        )
+      }
+    }
+
+    for await _ in stream {
+      // This for-await loop will exit (and thus `connect(lazily:)` will return)
+      // only when the task is cancelled, or when the stream's continuation is
+      // finished - whichever happens first.
+      // The continuation will be finished when `close()` is called and there
+      // are no more open streams.
+    }
+  }
+
+  /// Signal to the transport that no new streams may be created.
+  ///
+  /// Existing streams may run to completion naturally but calling ``withStream(descriptor:_:)``
+  /// should result in an ``RPCError`` with code ``RPCError/Code/failedPrecondition`` being thrown.
```

Comment:
We're implementing this here so we can be more concrete: 

```suggestion
  /// will result in an ``RPCError`` with code ``RPCError/Code/failedPrecondition`` being thrown.
```